components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/threatprotection/AnalyzerHolder.java

```diff
@@ -0,0 +1,108 @@
+
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.carbon.apimgt.gateway.threatprotection.analyzer.APIMThreatAnalyzer;
+import org.wso2.carbon.apimgt.gateway.threatprotection.analyzer.JSONAnalyzer;
+import org.wso2.carbon.apimgt.gateway.threatprotection.analyzer.XMLAnalyzer;
+import org.wso2.carbon.apimgt.gateway.threatprotection.configuration.ConfigurationHolder;
+import org.wso2.carbon.apimgt.gateway.threatprotection.configuration.JSONConfig;
+import org.wso2.carbon.apimgt.gateway.threatprotection.configuration.XMLConfig;
+import org.wso2.carbon.apimgt.gateway.threatprotection.pool.AnalyzerPool;
+import org.wso2.carbon.apimgt.gateway.threatprotection.pool.JSONAnalyzerFactory;
+import org.wso2.carbon.apimgt.gateway.threatprotection.pool.XMLAnalyzerFactory;
+import org.wso2.carbon.apimgt.gateway.threatprotection.utils.ThreatProtectorConstants;
+
+/**
+ * Holds the object pools for XML and JSON Analyzers
+ */
+public class AnalyzerHolder {
+
+    private static Logger log = LoggerFactory.getLogger(AnalyzerHolder.class);
+    private static AnalyzerHolder instance = new AnalyzerHolder();
+    private static AnalyzerPool<XMLAnalyzer> xmlAnalyzerAnalyzerPool;
+    private static AnalyzerPool<JSONAnalyzer> jsonAnalyzerAnalyzerPool;
+    private static GenericObjectPoolConfig poolConfig;
+
+    static {
+        poolConfig = new GenericObjectPoolConfig();
+        poolConfig.setMaxTotal(200);
+
+        poolConfig.setBlockWhenExhausted(false);
+        poolConfig.setMaxWaitMillis(0);
+
+        xmlAnalyzerAnalyzerPool = new AnalyzerPool<>(new XMLAnalyzerFactory(), poolConfig);
+        jsonAnalyzerAnalyzerPool = new AnalyzerPool<>(new JSONAnalyzerFactory(), poolConfig);
+    }
+
+    private AnalyzerHolder() {
+    }
+
+    public static AnalyzerHolder getInstance() {
+        return instance;
+    }
+
+    /**
+     * Borrows an object from pools (xml or json) for threat analysis
+     *
+     * @param contentType Content-Type of the payload
+     * @return Instance of APIMThreatAnalyzer based on content type
+     */
+    public static APIMThreatAnalyzer getAnalyzer(String contentType) {
+        APIMThreatAnalyzer analyzer = null;
+        if (ThreatProtectorConstants.TEXT_XML.equalsIgnoreCase(contentType) ||
+                ThreatProtectorConstants.APPLICATION_XML.equalsIgnoreCase(contentType)) {
+            try {
+                analyzer = xmlAnalyzerAnalyzerPool.borrowObject();
+                XMLConfig xmlConfig = ConfigurationHolder.getXmlConfig();
+                analyzer.configure(xmlConfig);
+            } catch (Exception e) {
+                // here apache.commons GenericObjectPool's borrow object method throws generic exception.
+                // here log the stacktrace along with the message.
+                log.error("Threat Protection: Error occurred while getting an object from the pool.", e);
```

Comment:
With this exception, can we proceed the execution?

Reply:
here we can continue the flow.